awriter/writer.go

```diff
@@ -17,355 +17,207 @@ package awriter
-	if err != nil {
-		return nil, errors.Wrapf(err, "writer: error finding parser for [%v]", tInfo.Type)
+
+	s := artifact.NewChecksumStore()
+	// calculate checksums of all data files
+	// we need this regardless of which artifact version we are writing
+	if err := calcDataHash(s, upd); err != nil {
+		return err
 	}
 
-	data, err := getDataFiles(filepath.Join(dir, cur, "data"))
+	// write temporary header (we need to know the size before storing in tar)
+	tmpHdr, err := writeTempHeader(s, devices, name, upd)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	defer os.Remove(tmpHdr.Name())
 
-	upd := parser.UpdateData{
-		Path:      filepath.Join(dir, cur),
-		DataFiles: data,
-		Type:      tInfo.Type,
-		P:         p,
-	}
-	return &upd, nil
-}
+	// mender archive writer
+	tw := tar.NewWriter(aw.w)
+	defer tw.Close()
 
-func (av *Writer) ScanUpdateDirs(dir string) ([]parser.UpdateData, error) {
-	// first check  if we have update in current directory
-	upd, err := av.readDirContent(dir, "")
-	if err == nil {
-		return []parser.UpdateData{*upd}, nil
-	} else if err != os.ErrInvalid {
-		return nil, err
+	// write version file
+	inf := artifact.ToStream(&artifact.Info{Version: version, Format: format})
+	sa := artifact.NewTarWriterStream(tw)
+	if err := sa.Write(inf, "version"); err != nil {
+		return errors.Wrapf(err, "writer: can not write version tar header")
 	}
 
-	dirs, err := ioutil.ReadDir(dir)
-	if err != nil {
-		return nil, err
+	// add checksum of `version`
+	if aw.signer != nil {
+		ch := artifact.NewWriterChecksum(ioutil.Discard)
+		ch.Write(inf)
+		s.Add("version", ch.Checksum())
 	}
 
-	updates := make([]parser.UpdateData, 0, len(dirs))
-	for _, uDir := range dirs {
-		if uDir.IsDir() {
-			upd, err := av.readDirContent(dir, uDir.Name())
-			if err == os.ErrInvalid {
-				continue
-			} else if err != nil {
-				return nil, err
-			}
-			updates = append(updates, *upd)
+	switch version {
+	case 2:
+		// write `manifest` file
+		sw := artifact.NewTarWriterStream(tw)
+		if err := sw.Write(s.GetRaw(), "manifest"); err != nil {
+			return errors.Wrapf(err, "writer: can not write manifest stream")
 		}
-	}
-
-	if len(updates) == 0 {
-		return nil, errors.New("writer: no update data detected")
-	}
-	return updates, nil
-}
 
-func (h *aHeader) closeHeader() (err error) {
-	// We have seen some of header components to cause crash while
-	// closing. That's why we are trying to close and clean up as much
-	// as possible here and recover() if crash happens.
-	defer func() {
-		if r := recover(); r != nil {
-			err = errors.New("error closing header")
+		// write signature
+		if err := writeSignature(tw, s.GetRaw(), aw.signer); err != nil {
+			return err
 		}
-		if err == nil {
-			h.isClosed = true
-		}
-	}()
+		// write header
 
-	if !h.isClosed {
-		errArch := h.hArchiver.Close()
-		errComp := h.hCompressor.Close()
-		errFile := h.hTmpFile.Close()
+	case 1:
+		// write header later on
```

Comment:
`write header` and `write header later on`? I'm confused by these comments, it looks like they take the same code path.